### PR TITLE
clean up raycluster instances at chart uninstall

### PR DESCRIPTION
**Pull request by akihikokuroda**
### Summary
fix: #695 

### Details and comments
Add pre-delete hook that deletes all remaining raycluster instances at helm uninstall.

### CODE REVIEW

**Comment by Tansito** [comment]
This is quite interesting. Do we want to remove all `ray-clusters` in `pre-install`? Thinking that experiments usually takes a lot of time and we will find situations where we should not delete a `cluster` from a client. Am I missing something?

**Comment by akihikokuroda** [comment]
We install and delete the Ray operator that creates and deletes that ray cluster based on the raycluster CRD instances.  We don't allow the others to manipulate the instances in the namespace.  So I think that it's safe and good to delete all in the namespace.

**Comment by akihikokuroda** [comment]
The helm has different hooks for the upgrade.  The pre-delete hook is executed only at uninstall.

**Comment by psschwei** [comment]
@Tansito @akihikokuroda  -- where did we land on merging this?

**Comment by akihikokuroda** [comment]
So far this doesn't get consensus so I can close or I can update to delete the instances only has the label.

**Comment by psschwei** [comment]
I recall running into issues trying to uninstall and reinstall the Helm chart (for local dev, I usually just kill the cluster and start over if I need to redeploy), so I'd be interested in seeing it merge.

**Comment by akihikokuroda** [comment]
I'll work to change only remove the labeled instance. @Tansito What do you think?

**Comment by psschwei** [comment]
> I'll work to change only remove the labeled instance. @Tansito What do you think?

My preference would be to just delete everything -- if we're uninstalling the Helm chart, I would think we want to clean up everything that's out there and not leave any orphans hanging. And if the pre-delete hook only gets called if we uninstall (and not when we upgrade), then I _think_ we should be ok.

**Comment by akihikokuroda** [comment]
@psschwei OK.  The helm chart install / uninstall the CRD so I agree it's OK to delete everything at uninstall.  I'll make the change.

**Review by psschwei — COMMENTED** [review]
LGTM

/assign @Tansito

**Comment by Tansito** [comment]
I'm back! As soon as I finish to read everything I will start reviewing this, thanks @akihikokuroda and @psschwei as always ❤️

**Comment by Tansito** [comment]
@akihikokuroda pffff sorry, I left some comments `pending` a lot of days ago but I didn't submit them (I don't know what happened 🤷‍♂️ )

File: infrastructure/helm/quantumserverless/templates/predelete.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      name: "{{ .Release.Name }}"
+      labels:
+        app.kubernetes.io/managed-by: {{ .Release.Service | quote }}
+        app.kubernetes.io/instance: {{ .Release.Name | quote }}
+        helm.sh/chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    spec:
+      restartPolicy: Never
+      serviceAccountName: gateway
+      containers:
+      - name: pre-install-job
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply infrastructure/helm/quantumserverless/templates/predelete.yaml]: I might copy it from somewhere :-(. I'll change it.

**Review by psschwei — APPROVED** [review]
LGTM

**Review by Tansito — APPROVED** [review]
Now it's fine for me! Sorry for the confusion 😂

**Comment by akihikokuroda** [comment]
@Tansito NP.  Thanks!

**Pull request merged**